MuteToggler@jebeaudet.com/files/MuteToggler@jebeaudet.com/applet.js

```diff
@@ -37,12 +37,29 @@ MyApplet.prototype = {
             if (this.settings) {
                 this.on_settings_changed();
             }
+
+            this.amixer = "";
+            const parameters = ["", " -D pulse"];
+            for (let param of parameters) {
+                let cmd = "amixer" + param + " scontrols";
+                global.log("MuteToggler: Test mixer command '" + cmd + "'");
+                let [res, stdout] = GLib.spawn_command_line_sync(cmd);
+                var string = new TextDecoder().decode(stdout);
+                if (res && string.indexOf("'Capture'") != -1) {
+                    this.amixer = "amixer" + param;
+                    global.log("MuteToggler: Use mixer command '" + this.amixer + "'"
+                    break;
+                } 
+            }
+            if (this.amixer == "") {
+                throw "No working mixer command found!";
```

Comment:
exceptions are catched lower and we'll end up with an invalid command "", could you log an error and fallback on the default `amixer` command at least?

Reply:
I see. I have a built in microphone, so never thought about this.

What do you think about not using this logic during start-up, but whenever the microphone is toggled? That should take care of hot-plugged microphones.

Reply:
Or maybe even easier: Use the command `sh -c "amixer set Capture toggle || amixer -D pulse set Capture toggle"`? This works, because if `amixer` cannot find the the device, its returned error code is >0.

Reply:
~Or maybe even easier: Use the command `sh -c "amixer set Capture toggle || amixer -D pulse set Capture toggle"`? This works, because if `amixer` cannot find the the device, its returned error code is >0.~
I just tried to implement this, but unfortunately I could not get it to work in the applet code (it works on the terminal).

Reply:
IMHO there are three sides to this:
1. There is the configurable hotkey for muting the microphone in the system setting. This functionality is the same as that for the MuteToggler hotkeys.
2. The sound applet already have a middle-click functionality. It can be used to toggle the microphone (and some other things). So this functionality is already there (but with the middle click instead of clicking onto MuteToggler). I just added a shift-middle click, so that you can toggle play _and_  (separately) toggle the microphone.
3. The MuteToggler shows the current status of the microphone. This is not shown in the standard sound applet. This is the only missing feature so that the sound applet can take over all functionality of the MuteToggler.

The standard sound applet uses the standard icons. I could not find an icon family which shows the output and input state in one icon. I guess, an alternative approach is needed, such as overlaying a badge onto an existing icon. Do you happen to know, how this can be done?

Reply:
Glad to hear that it's now all built in in the sound applet, can't remember if it was at the time I did this. But yea, a mute indicator would be the best and could put the final nail in the coffin for this applet.

I unfortunately have no idea how to do a badge. Reverse engineering the sound applet should not be too bad if you want to try it, IIRC, the sound applet icon showed a red X when the sound output is muted no? You could try to find this behavior in the codebase and see how it's done to mimic the same thing for the current active mic input.

Reply:
I just opened a discussion (https://github.com/orgs/linuxmint/discussions/209).

With regard to the badges: [SpicesUpdate@claudiux](https://github.com/linuxmint/cinnamon-spices-applets/tree/master/SpicesUpdate@claudiux) uses text badges implemented as [Label](https://gjs-docs.gnome.org/st13~13-label/) on its icon. Maybe it's easy to replace the label with an [Icon](https://gjs-docs.gnome.org/st13~13-icon/)?